### PR TITLE
feat: add initial implementation for to_timezone

### LINKED ISSUE #2029: function: to_timezone(timestamp, timezone)
### What type of enhancement is this?

API improvement

### What does the enhancement do?

We need a function to cast a timestamp into other timezones such as:

```sql
SELECT to_timezone(1623167145000000, 'Europe/Berlin')
```

### Implementation challenges

_No response_

### PR DESCRIPTION
**Pull request by parkma99**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

The feature adds a new UDF `to_timezone(timestamp, timezone)`

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)
closes #2029

### CODE REVIEW

File: tests/cases/standalone/common/function/time.sql [hunk 1]
```diff
@@ -1,3 +1,5 @@
 -- SQLNESS REPLACE (\d+:\d+:\d+\.\d+) TIME
 -- SQLNESS REPLACE [\s\-]+
 select current_time();
+
+SELECT to_timezone('2022-09-20T14:16:43.012345+08:00', 'Europe/Berlin');
```
Comment (parkma99) [inline tests/cases/standalone/common/function/time.sql:5]: I am confusing how to pass int64 as argument?, I passed a number, got `StringType`?

File: src/common/function/src/scalars/timestamp/to_timezone.rs [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+        .collect::<Vec<Option<String>>>()
+}
+
+impl Function for ToTimeZoneFunction {
+    fn name(&self) -> &str {
+        NAME
+    }
+
+    fn return_type(&self, _input_types: &[ConcreteDataType]) -> Result<ConcreteDataType> {
+        Ok(ConcreteDataType::string_datatype())
```
Comment (parkma99) [inline src/common/function/src/scalars/timestamp/to_timezone.rs:68]: I am not sure the return_type?

File: src/common/function/src/scalars/timestamp/to_timezone.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        let times = match columns[0].data_type() {
+            ConcreteDataType::String(_) => {
+                let array = columns[0].to_arrow_array();
+                let vector = StringVector::try_from_arrow_array(&array).unwrap();
+                Ok((0..vector.len())
+                    .map(|i| convert_to_timestrap(&vector.get(i).to_string()))
+                    .collect::<Vec<_>>())
+            }
+            _ => UnsupportedInputDataTypeSnafu {
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/common/function/src/scalars/timestamp/to_timezone.rs:102]: You can use `cast`:

```sql
select to_timezone('2023-07-31'::timestamp', 'utc');
```

We can always use `data::{type}` to cast the data into the special type,it's the same as `cast(data as {type})`.

File: tests/cases/standalone/common/function/time.sql (hunk 1 above)
Comment (killme2008) [reply tests/cases/standalone/common/function/time.sql:5]: The sqlness will render the timestamp in the string form from the results.

File: src/common/function/src/scalars/timestamp/to_timezone.rs (hunk 2 above)
Comment (parkma99) [reply src/common/function/src/scalars/timestamp/to_timezone.rs:68]: https://github.com/GreptimeTeam/greptimedb/blob/5bd80a74abe1f16a769e6298aaad4d64347a6464/src/common/time/src/timestamp.rs#L34-L38

Does timestamp member `value` mean the `$value * $unit` after the ` 1 January 1970`??
If so, I think the value should be unchanged, and return the `time string` with `local time zone`.

File: src/common/function/src/scalars/timestamp/to_timezone.rs [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+            }
+            match v {
+                Value::String(time) => {
+                    assert_eq!(time.as_utf8(), (*results.get(i).unwrap()).unwrap());
+                }
+                _ => unreachable!(),
+            }
+        }
+    }
+    fn build_vector_from_slice<T: ScalarVector>(items: &[Option<T::RefItem<'_>>]) -> T {
```
Comment (evenyag) [inline src/common/function/src/scalars/timestamp/to_timezone.rs:492]: You could use `from_owned_iterator()`
https://github.com/GreptimeTeam/greptimedb/blob/6d64e1c29639a4a079f8dd5bb5252912e537200f/src/datatypes/src/scalars.rs#L110

File: src/common/function/src/scalars/timestamp/to_timezone.rs (hunk 2 above)
Comment (evenyag) [reply src/common/function/src/scalars/timestamp/to_timezone.rs:78]: The timestamp is always the offset after `1970-1-1`. Timestamp itself isn't timezone aware so there is no need to convert a timestamp to a timestamp. I think this function only needs to format the timestamp with a specific timezone.

So it isn't as powerful as PostgresSQL's `timestamptz` type.
https://stackoverflow.com/questions/5876218/difference-between-timestamps-with-without-time-zone-in-postgresql
```sql
foo=> SET TIMEZONE TO 'Australia/Melbourne';
SET
foo=> SELECT '2011-01-01 00:00:00'::TIMESTAMP;
      timestamp      
---------------------
 2011-01-01 00:00:00
(1 row)
```

PostgreSQL also provides a `timezone()` function but it converts a `text` to a timestamp according to time zone. Actually.

A more flexible way might be to allow attaching the timezone info to our vector's data type like arrow:
https://docs.rs/arrow/latest/arrow/array/struct.PrimitiveArray.html#method.with_timezone

**Review by evenyag — COMMENTED** [review]
I think we might need a timestamptz type or attach timezone to the timestamp data type in the future. It is more powerful.

**Comment by parkma99** [comment]
Thanks @evenyag, This week I am on a business trip， I will handle it in weekend.

**Comment by parkma99** [comment]
> A more flexible way might be to allow attaching the timezone info to our vector's data type like arrow:
https://docs.rs/arrow/latest/arrow/array/struct.PrimitiveArray.html#method.with_timezone

How to do it ??

**Comment by evenyag** [comment]
> > A more flexible way might be to allow attaching the timezone info to our vector's data type like arrow:
> > https://docs.rs/arrow/latest/arrow/array/struct.PrimitiveArray.html#method.with_timezone
> 
> How to do it ??

Our primitive vectors are similar to arrow's primitive arrays. So we might use the same method to do this.

We need to implement a trait like [ArrowTimestampType](https://docs.rs/arrow/latest/arrow/datatypes/trait.ArrowTimestampType.html#)
https://github.com/apache/arrow-rs/blob/90449ffb2ea6ceef43ce8fc97084b3373975f357/arrow-array/src/types.rs#L309-L310

e.g.
`trait LogicalTimestampType: LogicalPrimitiveType {}`
https://github.com/GreptimeTeam/greptimedb/blob/develop/src/datatypes/src/types/primitive_type.rs#L80-L81

And then we can implement a `with_timezone` for our timestamp vectors like what arrow does:
https://github.com/apache/arrow-rs/blob/90449ffb2ea6ceef43ce8fc97084b3373975f357/arrow-array/src/array/primitive_array.rs#L1366-L1369

But we might need to do it in another PR.

**Review by evenyag — APPROVED** [review]
Rest LGTM. Please fix clippy warnings @parkma99

**Comment by parkma99** [comment]
Thanks for review @evenyag 😄

**Comment by evenyag** [comment]
Hi, we are going to switch to the `main` branch, as mentioned in https://github.com/GreptimeTeam/greptimedb/issues/3025.

I have changed the base branch of this PR to `main`.

File: src/common/function/src/scalars/timestamp/to_timezone.rs [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        })
+        .collect::<Vec<Option<String>>>()
+}
+
+impl Function for ToTimeZoneFunction {
+    fn name(&self) -> &str {
+        NAME
+    }
+
+    fn return_type(&self, _input_types: &[ConcreteDataType]) -> Result<ConcreteDataType> {
```
Comment (tisonkun) [inline src/common/function/src/scalars/timestamp/to_timezone.rs:77]: @evenyag @killme2008 if we choose the timestamp type as return type, which one should be returned here:

```rust
pub enum TimestampType {
    Second(TimestampSecondType),
    Millisecond(TimestampMillisecondType),
    Microsecond(TimestampMicrosecondType),
    Nanosecond(TimestampNanosecondType),
}
```

The input type can be string or integer other than timestamp so we cannot infer the type the same as the input timestamp type.

Comment (evenyag) [reply src/common/function/src/scalars/timestamp/to_timezone.rs:77]: Maybe return an error if input types are not timestamps? We should return a timestamp with the same time unit as the input one.

Comment (tisonkun) [reply src/common/function/src/scalars/timestamp/to_timezone.rs:77]: This makes sense. But how if users want to `to_timezone(ts_str, tzname)`? They should cast the string to timestamp themselves? And how.

**Comment by tisonkun** [comment]
I create a new PR based on this at https://github.com/GreptimeTeam/greptimedb/pull/3470.

**Pull request closed without merging**